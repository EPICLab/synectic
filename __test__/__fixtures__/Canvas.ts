import type { Canvas } from '../../src/types';
import { testStore } from './ReduxStore';

export const fullCanvas: Canvas = testStore.canvas;
