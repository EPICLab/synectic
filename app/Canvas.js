"use strict";
const Error = require('../lib/error.js');
const uuidv4 = require('uuid/v4');
const Dialog = require('../app/Dialog.js');
const Card = require('../app/Card.js');

module.exports = class Canvas {
  constructor({
    id = Error.throwIfMissing('id')
  }) {
    this.id = id;
    this.uuid = uuidv4();
    this.cards = [];

    this.canvas = document.createElement('div');
    $(this.canvas).attr('class', 'canvas');

    const versionDialogButton = document.createElement('button');
    $(versionDialogButton).click(() => { this.displayVersion(); });
    $(versionDialogButton).html('Version');

    const modalDialogButton = document.createElement('button');
    $(modalDialogButton).click(() => { this.addObj(); });
    $(modalDialogButton).html('Modal Dialog');

    const addCardButton = document.createElement('button');
    $(addCardButton).click(() => { this.addCard('text'); });
    $(addCardButton).html('Add Card');

    const printCardsButton = document.createElement('button');
    $(printCardsButton).click(() => { this.printCards(); });
    $(printCardsButton).html("Print Card(s)");

    //allow selection of cards using a lasso
    $(this.canvas).selectable({
      filter: 'div.card',
      cancel: 'input,textarea,button,select,option',
      selecting: function (event, ui) {
        $(ui.selecting).addClass('highlight');
      },
      unselecting: function (event, ui) {
        $(ui.unselecting).removeClass('highlight');
      },
      selected: function (event, ui) {
        $(ui.selected).addClass('highlight');
        window.addEventListener('contextmenu', (e) => {
          e.preventDefault()
          let context_menu = AppManager.rectangleMenuBuilder()
          context_menu.popup()
        }, false)
      },
      unselected: function (event, ui) {
        $(ui.unselected).removeClass('highlight');
      }
    });



    //allow the usage of the Meta (Ctrl/Command) key to select/deselect
    //single cards
    $( ".selector" ).bind( "mousedown", function ( e ) {
      e.metaKey = true;
    } ).selectable();

    this.canvas.appendChild(versionDialogButton);
    this.canvas.appendChild(modalDialogButton);
    this.canvas.appendChild(document.createElement('br'));
    this.canvas.appendChild(addCardButton);
    this.canvas.appendChild(printCardsButton);
    document.body.appendChild(this.canvas);
  }

  addCard(cardType = 'text', modality = true) {
    let card = new Card({
      id: this.nextCardId(),
      type: cardType,
      context: this.canvas,
      modal: modality
    });
    this.cards.push(card);
    return card;
  }

  removeCard({id, uuid} = {}) {
    if (uuid !== undefined) {
      let found = this.cards.find(card => card.uuid === uuid);
      let index = this.cards.indexOf(found);
      found.destructor();
      this.cards.splice(index, 1);
    } else if (id !== undefined) {
      let found = this.cards.find(card => card.id === id);
      let index = this.cards.indexOf(found);
      found.destructor();
      this.cards.splice(index, 1);
    } else {
      Error.throwIfMissingMinimum(1, 'id', 'uuid');
    }
  }

  nextCardId() {
    let ids = this.cards.map((card) => {
      return card.id;
    });
    if (ids.length < 1) return 1; // no cards on canvas yet

    let next = 1;
    while (ids.indexOf(next += 1) > -1);
    return next;
  }

  printCards() {
    console.log('CARDS (' + this.cards.length + ')');
    for (var card of this.cards) {
      card.printMetadata();
    }
  }

  addObj() {
    let dialog = Dialog.dialog({context: this.canvas});

    const ackButton = document.createElement('button');
    $(ackButton).attr('class', 'acknowledge');
    $(ackButton).html('OK');
    $(ackButton).click(function() { this.closest('.dialog').remove(); });
    dialog.appendChild(ackButton);

    let myNotification = new Notification('Title', {
      body: 'username: ' + require('username').sync()
    });

    myNotification.onclick = () => {
      console.log('Notification clicked')
    };
  };

  displayVersion() {
    let dialog = Dialog.notice({height: '300px', width: '400px', context: this.canvas});
    let remoteApp = require('electron').remote.app;

    const logo = document.createElement('img');
    const versionText = document.createElement('div');

    $(logo).attr('id', 'version_logo');
    $(versionText).attr('id', 'version_text');
    $(versionText).html(
      remoteApp.getName() + ' ' + remoteApp.getVersion() + '<br/>' +
      '<span id=\'frameworks\'>' +
      'Node.js ' + process.versions.node + ', ' +
      'Chrome ' + process.versions.chrome + ', and ' +
      'Electron ' + process.versions.electron + '</span>'
    );

    dialog.appendChild(logo);
    dialog.appendChild(versionText);
  }
};
