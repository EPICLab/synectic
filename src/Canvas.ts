import { Base } from './Base';

export class Canvas extends Base {

  constructor(children?: Base[]) {
    super(undefined, children);
    this.element.setAttribute('class', 'canvas');
    
  }
}
