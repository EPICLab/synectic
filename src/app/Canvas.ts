import { v4 } from 'uuid';
import * as $ from 'jquery';

export class Canvas {

  public readonly uuid: string = v4();
  public base: HTMLDivElement;

  constructor() {
    this.base = document.createElement('div');
    $(this.base).attr({class: 'canvas'});
    document.body.appendChild(this.base);
  }

}
