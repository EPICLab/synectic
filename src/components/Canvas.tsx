import './style.css';
import React, { useState, ReactNode } from 'react';
import Button from '@material-ui/core/Button';
import Card, { CardProps } from './Card';

export type CanvasProps = {
  children?: ReactNode;
}

const names = ['Bob', 'Sally', 'Billy', 'Jeanette', 'Ralph'];

const Canvas: React.FunctionComponent<CanvasProps> = (props: CanvasProps) => {
  const [cards, setCards] = useState<CardProps[]>([{ id: 1, name: 'Henry', offset: 0 }]);

  function createNewCard() {
    const newCards = [...cards];
    const newCard: CardProps = { id: cards.length + 1, name: names[cards.length], offset: (cards.length + 1) * 10 };
    newCards.push(newCard);
    setCards(newCards);
  }

  return (
    <div className='canvas' >
      <Button variant="contained" color="primary" onClick={() => createNewCard()}>New Card...</Button>
      <Button variant="contained" color="primary" onClick={() => console.log(cards)}>Display Cards...</Button>
      {cards.map(c => {
        return (<Card key={c.id} id={c.id} name={c.name} offset={c.offset} />);
      })}
      {props.children}
    </div>
  );
}

export default Canvas;







// // import React, { useState } from 'react';
// import React from 'react';

// interface Canvas {
//   timerID: NodeJS.Timeout | undefined;
//   date: Date;
//   counter: number;
// }


// export class CanvasComponent extends React.Component<Canvas, {}> {

//   constructor(props: Canvas) {
//     super(props);
//     this.state = { date: new Date(), counter: 0 };
//   }

//   componentDidMount() {
//     this.timerID = setInterval(() => this.tick(), 1000);
//   }

//   componentWillUnmount() {
//     if (this.timerID) clearInterval(this.timerID);
//   }

//   tick() {
//     this.setState({ date: new Date() });
//   }

//   click() {
//     this.setState((state: Readonly<{ date: Date; counter: number }>, props: Readonly<{ increment: number }>) => ({
//       counter: state.counter + props.increment
//     }));
//   }

//   render() {
//     return (
//       <div>
//         <h1>Hello, world!</h1>
//         <h2>It is {this.state.date.toLocaleTimeString()}.</h2>
//       </div>
//     );
//   }
// }




// // export function Canvas(): JSX.Element {
// //   const [count, setCount] = useState(0);

// //   return (
// //     <div>
// //       <p>You clicked {count} times</p>
// //       <button onClick={() => setCount(count + 1)}>
// //         Click me
// //       </button>
// //     </div>
// //   )
// // }