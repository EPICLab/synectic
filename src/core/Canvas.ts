import { Base } from './Base';

export class Canvas extends Base {

  constructor(children?: Base[]) {
    super(undefined, children);
    this.element.setAttribute('class', 'canvas');
    document.body.appendChild(this.element);

    document.addEventListener('remove', (e) => {
      let object: Base = this.find((e as CustomEvent).detail)[0];
      console.log('EVENT: remove ' + object.uuid);
      this.remove(object);
    }, false);
  }

  public add<T extends Base>(object: T): void {
    this.children.push(object);
    this.element.appendChild(object.element);
  }

  public remove<T extends Base>(object: T): boolean {
    if (this.children.some(c => c === object)) {
      this.children = this.children.filter(c => c !== object);
      return true;
    } else {
      return false;
    }
  }

  public find(uuid: string): Base[] {
    return this.children.filter(c => c.uuid === uuid);
  }
}
