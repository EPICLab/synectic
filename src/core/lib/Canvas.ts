import { Base } from './base';
import { Stack } from './Stack';
import { Card } from './Card';
import { v4 } from 'uuid';
import { DateTime } from 'luxon';

/**
 * Definition of a canvas to contain all open elements in context.
 */
export class Canvas implements Base<null, (Stack | Card)> {

  uuid: string = v4();
  created: DateTime = DateTime.local();
  modified: DateTime = DateTime.local();
  element: HTMLDivElement = document.createElement('div');
  parent: null;
  children: (Stack | Card)[] = [];

  /**
   * Default constructor for creating a canvas.
   * @param children Array of stacks and cards to populate the new canvas; can be empty.
   */
  constructor(children: (Stack | Card)[]) {
    children.map(c => this.add(c));
    this.element.setAttribute('class', 'canvas');
    document.body.appendChild(this.element);
  }

  /**
   * Default destructor for detaching from HTML DOM and deleting instance.
  */
  destructor(): void {
    document.body.removeChild(this.element);
    delete this.element;
  }

  /**
   * Adds stack or card to this canvas.
   * @param child A stack or card to be added.
   * @return Boolean indicating insertion success; false indicates child already exists in canvas.
  */
  add(child: Stack | Card): boolean {
    if (this.children.some(c => c.uuid === child.uuid)) {
      return false;
    } else {
      if (child instanceof Stack) child.parent.remove(child);
      if (child instanceof Card && child.parent instanceof Canvas) {
        child.parent.remove(child);
      }
      if (child instanceof Card && child.parent instanceof Stack) {
        child.parent.remove(child);
      }
      this.children.push(child);
      this.element.appendChild(child.element);
      return true;
    }
  }

  /**
   * Removes a stack or card from this canvas.
   * @param child A stack or card to be removed.
   * @return Boolean indicating removal success; false indicates child not found in canvas.
  */
  remove(child: Stack | Card): boolean {
    if (this.children.some(c => c.uuid === child.uuid)) {
      this.children = this.children.filter(c => c !== child);
      return true;
    } else {
      return false;
    }
  }

  /**
   * Search for a stack or card on this canvas.
   * @param uuid The unique user ID to search for within the canvas.
   * @return An array of stacks and cards that match the given uuid; can be empty.
  */
  search(uuid: string): (Stack | Card)[] {
    return this.children.filter(c => c.uuid === uuid);
  }
}
