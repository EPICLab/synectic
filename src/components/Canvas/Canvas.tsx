import { useDroppable } from '@dnd-kit/core';
import { DarkMode, LightMode } from '@mui/icons-material';
import { styled } from '@mui/material';
import { DateTime } from 'luxon';
import React, { PropsWithChildren, useContext } from 'react';
import { ErrorBoundary } from 'react-error-boundary';
import version from '../../../version';
import { fileOpenDialog } from '../../containers/dialogs';
import { useAppDispatch, useAppSelector } from '../../store/hooks';
import branchSelectors from '../../store/selectors/branches';
import cardSelectors from '../../store/selectors/cards';
import commitSelectors from '../../store/selectors/commits';
import filetypeSelectors from '../../store/selectors/filetypes';
import metafileSelectors from '../../store/selectors/metafiles';
import modalSelectors from '../../store/selectors/modals';
import repoSelectors from '../../store/selectors/repos';
import stackSelectors from '../../store/selectors/stacks';
import { modalAdded } from '../../store/slices/modals';
import { AppThemeContext } from '../AppTheme';
import Card from '../Card';
import DndCanvasContext from '../Dnd/DndCanvasContext';
import GitGraphSelect from '../GitGraph/GitGraphSelect';
import ModalComponent from '../Modal';
import NavMenu, { NavItemProps } from '../NavMenu';
import Stack from '../Stack';

const isMac = window.api.globals.platform === 'darwin';

const Canvas = ({ children }: PropsWithChildren) => {
  const metafiles = useAppSelector(state => metafileSelectors.selectAll(state));
  const cards = useAppSelector(state => cardSelectors.selectAll(state));
  const stacks = useAppSelector(state => stackSelectors.selectAll(state));
  const filetypes = useAppSelector(state => filetypeSelectors.selectAll(state));
  const modals = useAppSelector(state => modalSelectors.selectAll(state));
  const repos = useAppSelector(state => repoSelectors.selectAll(state));
  const branches = useAppSelector(state => branchSelectors.selectAll(state));
  const commits = useAppSelector(state => commitSelectors.selectAll(state));
  const dispatch = useAppDispatch();
  const { mode, toggleColorMode } = useContext(AppThemeContext);
  const { isOver, setNodeRef } = useDroppable({
    id: 'Canvas',
    data: {
      type: 'canvas'
    }
  });

  const fileMenu: NavItemProps[] = [
    {
      label: 'New...',
      click: () =>
        dispatch(
          modalAdded({
            id: window.api.uuid(),
            type: 'NewCardDialog'
          })
        )
    },
    ...(isMac
      ? [
          {
            label: 'Open...',
            click: () => dispatch(fileOpenDialog())
          }
        ]
      : [
          {
            label: 'Open File...',
            click: () => dispatch(fileOpenDialog({ properties: ['openFile', 'multiSelections'] }))
          },
          {
            label: 'Open Directory...',
            click: () =>
              dispatch(fileOpenDialog({ properties: ['openDirectory', 'multiSelections'] }))
          }
        ])
    // {
    //   label: 'Notify...',
    //   click: () => window.api.notifications.sendNotification('My custom message!')
    // },
    // {
    //   label: 'Context',
    //   click: async () => {
    //     console.log(`Executing in ${isRenderer()}`);
    //     const preloadContext = await window.api.context();
    //     console.log(`Preload in ${preloadContext}`);
    //   }
    // }
  ];

  const showStore = () => {
    console.group(
      `%cRedux Store : ${DateTime.local().toHTTP()}`,
      'background: lightblue; color: #444; padding: 3px; border-radius: 5px;'
    );
    console.log(`STACKS [${Object.keys(stacks).length}]`, stacks);
    console.log(`CARDS [${Object.keys(cards).length}]`, cards);
    console.log(`FILETYPES [${filetypes.length}]`, filetypes);
    console.log(`METAFILES [${metafiles.length}]`, metafiles);
    // console.log(`CACHE [${cache.length}]`, cache);
    console.log(`REPOS [${repos.length}]`, repos);
    console.log(`BRANCHES [${branches.length}]`, branches);
    console.log(`COMMITS [${commits.length}]`, commits);
    console.log(`MODALS [${modals.length}]`, modals);
    console.groupEnd();
  };

  const sysMenu: NavItemProps[] = [
    { label: 'View Datastore...', click: () => showStore() },
    {
      label: mode === 'light' ? 'Dark mode' : 'Light mode',
      icon: mode === 'light' ? <DarkMode /> : <LightMode />,
      click: () => toggleColorMode()
    }
  ];

  const helpMenu: NavItemProps[] = [
    {
      label: 'Website...',
      click: async () => {
        window.api.openExternal('https://nomatic.dev/synectic');
      }
    },
    {
      label: 'Repository...',
      click: async () => {
        window.api.openExternal('https://github.com/EPICLab/synectic/');
      }
    },
    {
      label: 'Release Notes...',
      click: async () => {
        window.api.openExternal('https://github.com/EPICLab/synectic/releases');
      }
    },
    {
      label: 'View License...',
      click: async () => {
        window.api.openExternal(
          'https://github.com/EPICLab/synectic/blob/5ec51f6dc9dc857cae58c5253c3334c8f33a63c4/LICENSE'
        );
      }
    },
    {
      label: 'Version',
      click: () =>
        dispatch(
          modalAdded({
            id: window.api.uuid(),
            type: 'Notification',
            options: { message: `Synectic v${version}` }
          })
        )
    }
  ];

  return (
    <AppContainer>
      <NavBar>
        <NavMenu label="File" submenu={fileMenu} />
        <NavMenu label="System" submenu={sysMenu} />
        <NavMenu label="Help" submenu={helpMenu} />
      </NavBar>
      <CanvasComponent id="Canvas" ref={setNodeRef} isOver={isOver}>
        <ErrorBoundary fallback={<Error>ðŸ’¥GitGraphSelect ErrorðŸ’¥</Error>}>
          <GitGraphSelect />
        </ErrorBoundary>
        <DndCanvasContext>
          <ErrorBoundary fallback={<Error>ðŸ’¥Stack ErrorðŸ’¥</Error>}>
            {stacks.map(stack => (
              <Stack key={stack.id} id={stack.id} />
            ))}
          </ErrorBoundary>
          <ErrorBoundary fallback={<Error>ðŸ’¥Card ErrorðŸ’¥</Error>}>
            {cards.map(card => (card.captured ? null : <Card key={card.id} id={card.id} />))}
          </ErrorBoundary>
          <ErrorBoundary fallback={<Error>ðŸ’¥Modal ErrorðŸ’¥</Error>}>
            {modals.map(modal => (
              <ModalComponent key={modal.id} {...modal} />
            ))}
          </ErrorBoundary>
          {children}
        </DndCanvasContext>
      </CanvasComponent>
    </AppContainer>
  );
};

const AppContainer = styled('div')(() => ({
  display: 'flex',
  flexFlow: 'column',
  height: '100vh'
}));

const NavBar = styled('div')(({ theme }) => ({
  padding: 1,
  backgroundColor: theme.palette.background.default,
  color: theme.palette.text.secondary
}));

const CanvasComponent = styled('div')<{ isOver: boolean }>(props => ({
  flex: 1,
  background: `url(${require('../../assets/canvas.png')}) center/auto fixed`,
  width: '100%',
  color: props.isOver ? 'green' : undefined
}));

const Error = styled('div')(() => ({
  color: '#FF0000'
}));

export default Canvas;
