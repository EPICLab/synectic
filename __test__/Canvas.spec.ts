import Canvas, { CanvasProps } from '../src/components/Canvas';

describe('Canvas', () => {
  const props: CanvasProps = {};

  it('creates an instance of Canvas', async () => {
    expect(Canvas(props)).toBeInstanceOf(Canvas);
  });

  it(`testing snapshots match`, () => {
    const bar = {
      foo: {
        x: 2,
        y: 2
      },
    };

    expect(bar).toMatchSnapshot();
  });
})