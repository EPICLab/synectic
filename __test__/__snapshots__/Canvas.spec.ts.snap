// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas testing snapshots match 1`] = `
Object {
  "foo": Object {
    "x": 2,
    "y": 2,
  },
}
`;
